### ISSUE TITLE
SDL_RenderClear not loaded

### ISSUE THREAD

**Issue by CKraniak** [FreeRCT/FreeRCT#131]
System: Windows 7, 32-bit (specifically, 32-bit build, 64-bit system / OS, at least I think I build 32-bit, hasn't been relevant for me yet)
Build system: MinGW ... not sure which one, since I think I have like 7 of them in my path, need to fix that

Make and Cmake run perfectly, except I have to pull out some quotes in the include_directories(libpng_dirs) in the CMakeLists.

The issue manifests after a successful build. Trying to run the program results in a segfault. I've been able to trace the problem to the call to SDL_RenderClear in VideoSystem::FinishRepaint(). For some reason, SDL_RenderClear seems to not load, so its address when it gets called is 0x0.

This is either a problem with the way I installed the library, the way my %Path% is so cluttered, something in SDL, or something in FreeRCT. I'm predisposed to thinking the %Path% is the problem. It could also be related to my system bluescreening twice, but I think that's the video driver getting confused about these libraries, or maybe the way I'm messing with %Path% isn't something some of the drivers can take.

I'm probably going to call LoadLibrary / GetProcAddress myself. Hopefully, I can at least get the game to open up.

---

As an aside: how is one supposed to converse with the other devs? I'm pretty new to Git / Github / open-source contribution, but I do know a little C++. I'm interested in creating substantial contributions (patrol system -> mechanic employees, maybe a "hill-climbing" pathfinder), but don't know how I'm supposed to converse about this.

**Comment by LordAro** [comment]
I would indeed recommend tidying up your %PATH% - will make working out what's going on much easier :)

Which modifications are you making to CMake wrt the quoted variables? I can take a look see what should and shouldn't be there

SDL_RenderClear not being defined is very odd and is not something I nor google seems to have heard of before. I'd suggest running with `make VERBOSE=1` to show exactly what's being built

As for conversing, here works fine, we're also hanging around on IRC - #freerct on OFTC.net

**Comment by CKraniak** [comment]
`include_directories($"{PNG_INCLUDE_DIR}")` becomes `include_directories(${PNG_INCLUDE_DIR})`. My Cmake spits out an include_CXX.rec (?) that says something like `-I<path/to/png.h/>;<path/to/zlib.h/>` and in the build ends up complaining about not being able to find them, apparently because it needed to be `-I<path/to/png.h/> -I<path/to/zlib.h/>` instead. It's trivial to fix, now that I know what it is, so I don't see a huge need to commit that if it breaks everyone else's build; I think I can live with manually changing that after merges and before commits.

According to SDL's website, version 2.0.3 was the last to <strike>officially support</strike> be known to build on MinGW32. I can either switch to MinGW64 or switch from SDL 2.0.4 to 2.0.3, iff FreeRCT doesn't depend on that specific version of SDL. Changing SDL would be easier, since otherwise I'd have to find a way to stop Cmake from building against one of the other MinGWs in my %Path%. But maybe that's easier than it sounds like: Qt has an option for which compiler you want to use, I just haven't used it before, since its default compiler is _normally_ fine. But Qt also seems to have a habit of getting out of Cmake's way, so maybe it won't be able to force Cmake to build against the right compiler. I don't know; I need to test it all later.

I don't know if that is the problem, but it could theoretically be. I'm still going to tidy up my path a bit later, pull perl out of it (since I believe one mingw is in perl's bin), pull Ruby out, maybe rebuild SDL, etc. And if none of that works, I'll probably see what happens if I manually get SDL_RenderClear's address. What's your policy on build-specific fixes? I was thinking about doing `#if defined(SDL_RENDERCLEAR_MANUAL_LOAD)`, probably right after SDL_Init, and passing a command line argument to set `SDL_RENDERCLEAR_MANUAL_LOAD` to Cmake. I don't see this addressed in the style guide. Thoughts?

(I guess I didn't mention I'm using Qt 5.5 as my editor. It doesn't seem important, since that's ... basically all it does in a Cmake project: code editor. Compiler is still MinGW, Cmake is Cmake, and I don't _think_ manually setting the build %Path% in Qt is actually affecting the %Path% Cmake sees, so it's just a code editor.)

**Comment by CKraniak** [comment]
Okay. So I now have a cleaner path and am pretty sure that's out of the way forever. Also built SDL, seems to be ok for now, although it didn't change anything.

New weirdness: when I try to add my modifications right before FinishPaint, the error moves to Blit8bppImages, which happens earlier in execution. It's happening when src_base gets dereferenced in the BlendPixels call; It looks like the blit_rect's memory is being overrun. 

I can't even get to the SDL_RenderClear to do anything if something else segfaults first.

I'll keep messing around with it.

**Comment by CKraniak** [comment]
Solved.

Explanation:

I have never used SDL before. I downloaded the runtime binaries / source code, since that's what it looked like you were supposed to do. But according to a tutorial I finally stumbled on [here](http://lazyfoo.net/tutorials/SDL/01_hello_SDL/windows/mingw/index.php), I was actually supposed to scroll just a bit further down and grab the "development libraries". Dropping those in and adjusting the path to point at those solved the issue.
